### PR TITLE
Loop-invariant code motion

### PR DESCRIPTION
**Pull request by abadams**
Lifts pure expressions out of inner loops. LLVM does this anyway, so it doesn't matter in most cases. However it's very useful for pulling expressions across device transitions, so that CUDA code (for example) doesn't redundantly redo a bunch of integer math per thread.

### CODE REVIEW

File: src/LICM.cpp [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    map<Expr, string, IRDeepCompare> lifted;
+};
+
+class LICM : public IRMutator {
+    using IRVisitor::visit;
+
+    bool in_gpu_loop {false};
+
+    void visit(const For *op) {
+        bool old_in_gpu_loop = in_gpu_loop;
```
Comment (steven-johnson) [inline src/LICM.cpp]: Y'know, we use this pattern pervasively enough in mutators that we can probably justify a templated helper for save_old+set_new+restore_old...

File: test/correctness/loop_invariant_extern_calls.cpp [hunk 2]
```diff
@@ -58,21 +64,38 @@ int main(int argc, char **argv) {
...
+    if (call_counter[3] != 1 || call_counter[4] != 32 * 32) {
+        printf("Call counter for parallel call was %d, %d instead of %d, %d\n",
+               call_counter[3], call_counter[4], 1, 32 * 32);
         return -1;
     }
 
+    // Check that something we can't compute on the GPU gets lifted
+    // out of the GPU loop. This code would fail to link if we didn't
+    // do loop invariant code motion.
+    if (get_target_from_environment().has_gpu_feature()) {
```
Comment (steven-johnson) [inline test/correctness/loop_invariant_extern_calls.cpp]: should this be get_jit_target_from_environment()?

File: src/LICM.cpp (hunk 1 above)
Comment (abadams) [reply src/LICM.cpp]: I haven't been able to come up with anything that actually reads as clearer, e.g:

```
class Foo : public IRMutator {
    MutatorState<bool> in_gpu_loop;

    void visit(const For *op) {
        auto old_in_gpu_loop = in_gpu_loop.push(op->for_type == ...);

        if (in_gpu_loop && !old_in_gpu_loop) {
            ...
        }
    
...

```

though automatically restoring the old value would be nice from a bug-avoidance perspective

Comment (steven-johnson) [reply src/LICM.cpp]:
```
class Foo : public IRMutator {
    void visit(const For *op) {
        SaveState<bool> in_gpu_loop(op->for_type == ...);
        if (in_gpu_loop && !in_gpu_loop.previous()) {
            ...
        }
    }
```

Comment (abadams) [reply src/LICM.cpp]: Doesn't there need to be a class member to actually hold the state?

Comment (steven-johnson) [reply src/LICM.cpp]: Yep. Should have proofread before posting.

**Comment by steven-johnson** [comment]
LGTM

**Pull request merged**